security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,428 @@
+          },
+          getStorageRootKey: (callback) => {
+            callback(null, this._srk.asTpm2B());
+          },
+          signWithIdentity: (deviceIdData, callback) => {
+            this._signData(deviceIdData, (errSign: Error, signedData: Buffer) => {
+              if (errSign) {
+                debug('Error from signing data: ' + errSign);
+                callback(errSign);
+                this._fsm.transition('disconnected');
+              } else {
+                callback(null, signedData);
+              }
+            });
+          },
+          activateSymmetricIdentity: (identityKey, callback) => {
+            this._activateSymetricIdentity(identityKey, (errActivate: Error) => {
+              if (errActivate) {
+                debug('Error from activate: ' + errActivate);
+                callback(errActivate);
+                this._fsm.transition('disconnected');
+              } else {
+                callback(null, null);
+              }
+            });
+          },
+        }
+      }
+    });
+  }
+
+  getEndorsementKey(callback: (err: Error, endorsementKey: string) => void): void {
+      this._fsm.handle('getEndorsementKey', callback);
+  }
+
+  getStorageRootKey(callback: (err: Error, storageKey: string) => void): void {
+    this._fsm.handle('getStorageRootKey', callback);
+  }
+
+  signWithIdentity(deviceIdData: Buffer, callback: (err: Error, signedResponse: Buffer) => void): void {
+    if (deviceIdData === null || deviceIdData.length === 0) {
+        throw new ReferenceError('\'deviceIdData\' cannot be \'' + deviceIdData + '\'');
+    }
+
+    if (this._idKeyPub == null) {
+        throw new errors.ArgumentError('activateIdentityKey first before signing');
+    }
+
+    this._fsm.handle('signWithIdentity', deviceIdData, callback);
   }
 
-  getStorageRootKey(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  activateSymmetricIdentity(identityKey: Buffer, callback: (err: Error, returnedActivate: Buffer) => void): void {
+    this._fsm.handle('activateSymmetricIdentiy', identityKey, callback);
   }
 
-  signWithIdentity(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  getRegistrationId(callback: (err: Error, registrationId: string) => void): void {
+    if (this._registrationId) {
+      callback(null, this._registrationId);
+    } else {
+      this.getEndorsementKey( function (endorsementError: Error, endorsementKey: string): void {
+        if (endorsementError) {
+          callback(endorsementError, null);
+        } else {
+          let hasher = crypto.createHash('sha256');
+          hasher.update(endorsementKey);
+          let hashedEndorsement: Buffer = hasher.digest();
+          let basedEndorsement: string = base32Encode(hashedEndorsement, 'RFC4648');
+          let loweredEndorsment: string = basedEndorsement.toLowerCase();
+          this._registrationId = loweredEndorsment.replace(/=/g, '');
```

Comment:
this might deserve a separate _createRegistrationIdFromEK method (single responsibility) and most of this variables are actually constants. (const instead of let)

Reply:
not sure why this method isn't part of the state machine too.

---
In reply to: [157890158](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r157890158) [](ancestors = 157890158)

Reply:
As for the state machine, I prefer to have fewer handlers.  This will of course activate the hash machine.

---
In reply to: [158117104](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r158117104) [](ancestors = 158117104,157890158)